external-crates/move/crates/move-compiler/src/naming/translate.rs

```diff
@@ -2245,54 +2280,95 @@ fn pat(context: &mut Context, sp!(ploc, pat_): E::MatchPattern) -> N::MatchPatte
 
     let pat_: N::MatchPattern_ = match pat_ {
         EP::PositionalConstructor(name, etys_opt, args) => {
-            if let Some((mident, enum_, variant, tys_opt, _, is_positional, is_empty)) =
+            if let Some((mident, enum_, variant, tys_opt, _, field_info)) =
                 context.resolve_variant_name(ploc, "pattern", name, etys_opt)
             {
-                if is_empty {
+                if field_info.is_empty() {
                     let msg = "Invalid variant pattern. Empty variants \
                                are not matched with positional variant syntax";
                     let mut diag = diag!(NameResolution::PositionalCallMismatch, (ploc, msg));
                     diag.add_note("Remove '()' after the variant name");
                     context.env.add_diag(diag);
-                } else if !is_positional {
+                } else if !field_info.is_positional() {
                     let msg = "Invalid variant pattern. Named variant declarations \
                                    require named patterns.";
                     context
                         .env
                         .add_diag(diag!(NameResolution::PositionalCallMismatch, (ploc, msg)));
                 }
 
-                let args = UniqueMap::maybe_from_iter(args.value.into_iter().enumerate().map(
-                    |(idx, p)| {
-                        let field = Field::add_loc(p.loc, format!("{idx}").into());
-                        (field, (idx, pat(context, p)))
-                    },
-                ))
-                .unwrap();
-                NP::Constructor(mident, enum_, variant, tys_opt, args)
+                let fields = field_info.field_count();
+
+                // NB: We may have more args than fields! Since we allow `..` to be zero-or-more
+                // wildcards.
+                let missing = (fields as isize) - args.value.len() as isize;
+                let mut idx = 0;
+                let mut result_args = UniqueMap::new();
+                for p in args.value.into_iter() {
+                    match p {
+                        E::Ellipsis::Ellipsis(loc) => {
+                            for _ in 0..=missing {
+                                let field = Field::add_loc(loc, format!("{idx}").into());
+                                result_args
+                                    .add(field, (idx, sp(loc, NP::Wildcard)))
+                                    .unwrap();
+                                idx += 1;
+                            }
+                        }
+                        E::Ellipsis::Binder(p) => {
+                            let field = Field::add_loc(p.loc, format!("{idx}").into());
+                            result_args.add(field, (idx, pat(context, p))).unwrap();
+                            idx += 1;
+                        }
+                    }
+                }
```

Comment:
Not that it particularly matters, but can you do this with take/drop?

```
let ellipsis_ndx = args.value.iter().position(|field| matches!(field, E::Ellipsis::Ellipsis(_)));
let args = if let Some(ndx) = ellipsis_ndx {
        let mut iter = args.value.into_iter();
        let mut args = iter.take(ndx);
        let Some(E::Ellipsis::Ellipsis(loc)) = iter.next() {
           ...
        }
        args.append(&mut iter.collect());
        args
    } else {
        args
    };
}
let args = UniqueMap::maybe_from_iter(args.value.into_iter().enumerate().map(
                    |(idx, p)| {
                        let field = Field::add_loc(p.loc, format!("{idx}").into());
                        (field, (idx, pat(context, p)))
                    },
                ))
                .unwrap();
```

Reply:
Didn't windup going with this, but it inspired me to go in this general direction and I like the way it turned out. Thanks for the suggestion!